report/report.go

```diff
@@ -469,6 +475,37 @@ func (r Report) upgradeNamespaces() Report {
 	return r
 }
 
+// Note in-place modification of Endpoint.Nodes
+func (r Report) upgradeDNSRecords() Report {
+	if len(r.DNS) > 0 {
+		return r
+	}
+	dns := make(DNSRecords)
+	for endpointID, endpoint := range r.Endpoint.Nodes {
+		_, addr, _, ok := ParseEndpointNodeID(endpointID)
+		snoopedNames, foundS := endpoint.Sets.Lookup(SnoopedDNSNames)
+		reverseNames, foundR := endpoint.Sets.Lookup(ReverseDNSNames)
+		if ok && (foundS || foundR) {
+			// Add address and names to report-level data - we assume
+			// all endpoints with the same address have equally good data
```

Comment:
I don't see why that assumption would hold. And if it doesn't then the data here does not converge, which results in view instability.

You could _optimise_ for the case that the assumption holds, but not depend on it.

Reply:
The data comes from the caches in `resolver.go` and `dns_snooper_linux.go`; if that data is changing then it's a matter of timing what any particular endpoint has.

I can't imagine why the data would change in such a way that results didn't converge shortly.

Reply:
AFAICT you could process the same report twice and get different results. That's a recipe for view instability.